### PR TITLE
[WIP] Initial HIP-ification of PyTorch code for AMD hardware

### PR DESCRIPTION
**Pull request by wsttiger**
This is the upstreaming effort by AMD to integrate hip into PyTorch. These modifications will allow continuous modifications to PyTorch code such that it will run on AMD GPU's. For now, this PR contains code that only runs on NVIDIA. There should be no side effects, i.e. everything should run normally on NV hardware. Please check ... thanks.

@soumith

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -19,7 +19,8 @@
 DEBUG = check_env_flag('DEBUG')
 WITH_DISTRIBUTED = not check_env_flag('NO_DISTRIBUTED')
 WITH_DISTRIBUTED_MW = WITH_DISTRIBUTED and check_env_flag('WITH_DISTRIBUTED_MW')
-WITH_NCCL = WITH_CUDA and platform.system() != 'Darwin'
+# WITH_NCCL = WITH_CUDA and platform.system() != 'Darwin'
+WITH_NCCL = False
```
Comment (fmassa) [inline setup.py]: Could we make it conditional on the compilation, so that we can still leverage NCCL on NVidia?

**Comment by wsttiger** [comment]
I’m sorry I left this commented.
> +# WITH_NCCL = WITH_CUDA and platform.system() != 'Darwin'
The compilation was failing both with the hip code and the master tip code when I had WITH_NCCL set to True.

-Scott

**Comment by soumith** [comment]
Scott, as we discussed IRL a couple of weeks ago, we'd want the PR to not introduce a hard-dependency on HIP.

1. The HIP calls/changes need to be guarded by `#ifdef` or `#if` macros
2. The changes have to be minimal. For example, I'd rather have the `hipify` regex run at build time, rather than pre-running it and checking in the changed code.

Overall, it should be able to build just fine using CUDA without any HIP stuff installed on my system whatsoever.

**Comment by akostadinov** [comment]
@soumith , excuse me if there have been other discussion that I'm not aware of. But how could pytorch project accept amd gpu support? On the one hand I understand the desire to still allow pure CUDA without any intermediate libraries. On the other hand doing `sed` during build time appears very fragile to me. Also I don't see how pure CUDA would ever support amd gpus. Then only other option seems to be to choose code path during build time but this would duplicate a lot of code increasing maintenance cost.

Again, please let me know if things have already been figured out. Would appreciate any thoughts you can provide.

**Comment by soumith** [comment]
@akostadinov a combination of `sed` + patch files during build time, and having continuous integration for AMD GPUs will help make this less fragile and manageable from a maintainer model.

From user perspective nothing changes, their code will work on AMD or CUDA GPUs the same without any `sed` or modification.

we are working closely with AMD engineers to get a PR opened that is more concise.

**Pull request closed without merging**